cmd/geth/admin.go

```diff
@@ -396,28 +400,243 @@ func (js *jsre) printBlock(call otto.FunctionCall) otto.Value {
 }
 
 func (js *jsre) dumpBlock(call otto.FunctionCall) otto.Value {
-	var block *types.Block
-	if len(call.ArgumentList) > 0 {
-		if call.Argument(0).IsNumber() {
-			num, _ := call.Argument(0).ToInteger()
-			block = js.ethereum.ChainManager().GetBlockByNumber(uint64(num))
-		} else if call.Argument(0).IsString() {
-			hash, _ := call.Argument(0).ToString()
-			block = js.ethereum.ChainManager().GetBlock(common.HexToHash(hash))
-		} else {
-			fmt.Println("invalid argument for dump. Either hex string or number")
-		}
-
-	} else {
-		block = js.ethereum.ChainManager().CurrentBlock()
-	}
-	if block == nil {
-		fmt.Println("block not found")
+	block, err := js.getBlock(call)
+	if err != nil {
+		fmt.Println(err)
 		return otto.UndefinedValue()
 	}
 
 	statedb := state.New(block.Root(), js.ethereum.StateDb())
 	dump := statedb.RawDump()
 	return js.re.ToVal(dump)
+}
 
+func (js *jsre) waitForBlocks(call otto.FunctionCall) otto.Value {
```

Comment:
This implements the `admin.debug.waitForBlocks` method so it is not called in go but on the console or in a js file (see link above).  

The mining example i show in the script was not possible since starting mining with the old big dag is not realistic in tests. This hack should be eliminated, once there is new ethash with minidag for tests. Then the xeth.ApplyTestTxs (used natspec and cmd/geth tests) should be replaced with a xeth.mineBlocks(n int)

As for the callback, I agree blocking the mainthread is incorrect IF we have real asynchrony, but until async send and watch are implemented we cannot have this as the single method offering  async callback. 
On the other hand, if async is properly implemented we can eliminate this method in favour of a filter:

```
eth.filter("latest", function(err, res) {
     if(num == 2)
           stopmining();
     num++;
})
```